File: distribution/python/dqops/client/models/rule_parameters_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.field_model import FieldModel


T = TypeVar("T", bound="RuleParametersModel")


@_attrs_define
class RuleParametersModel:
    """Model that returns the form definition and the form data to edit parameters (thresholds) for a rule at a single
    severity level (low, medium, high).

        Attributes:
            rule_name (Union[Unset, str]): Full rule name. This field is for information purposes and can be used to create
                additional custom checks that reuse the same data quality rule.
            rule_parameters (Union[Unset, List['FieldModel']]): List of fields for editing the rule parameters like
                thresholds.
            disabled (Union[Unset, bool]): Disable the rule. The rule will not be evaluated. The sensor will also not be
                executed if it has no enabled rules.
            configured (Union[Unset, bool]): Returns true when the rule is configured (is not null), so it should be shown
                in the UI as configured (having values).
    """

    rule_name: Union[Unset, str] = UNSET
    rule_parameters: Union[Unset, List["FieldModel"]] = UNSET
    disabled: Union[Unset, bool] = UNSET
    configured: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        rule_name = self.rule_name
        rule_parameters: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.rule_parameters, Unset):
            rule_parameters = []
            for rule_parameters_item_data in self.rule_parameters:
                rule_parameters_item = rule_parameters_item_data.to_dict()

                rule_parameters.append(rule_parameters_item)

        disabled = self.disabled
        configured = self.configured

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if rule_name is not UNSET:
            field_dict["rule_name"] = rule_name
        if rule_parameters is not UNSET:
            field_dict["rule_parameters"] = rule_parameters
        if disabled is not UNSET:
            field_dict["disabled"] = disabled
        if configured is not UNSET:
            field_dict["configured"] = configured

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.field_model import FieldModel

        d = src_dict.copy()
        rule_name = d.pop("rule_name", UNSET)

        rule_parameters = []
        _rule_parameters = d.pop("rule_parameters", UNSET)
        for rule_parameters_item_data in _rule_parameters or []:
            rule_parameters_item = FieldModel.from_dict(rule_parameters_item_data)

            rule_parameters.append(rule_parameters_item)

        disabled = d.pop("disabled", UNSET)

        configured = d.pop("configured", UNSET)

        rule_parameters_model = cls(
            rule_name=rule_name,
            rule_parameters=rule_parameters,
            disabled=disabled,
            configured=configured,
        )

        rule_parameters_model.additional_properties = d
        return rule_parameters_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/services/check/mapping/models/RuleParametersModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.services.check.mapping.models;

import com.dqops.rules.AbstractRuleParametersSpec;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Model that returns the form definition and the form data to edit parameters (thresholds) for a rule at a single severity level (low, medium, high).
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "RuleParametersModel", description = "Model that returns the form definition and the form data to edit parameters (thresholds) for a rule at a single severity level (low, medium, high).")
public class RuleParametersModel implements Cloneable {
    /**
     * Full rule name. This field is for information purposes and can be used to create additional custom checks that reuse the same data quality rule."
     */
    @JsonPropertyDescription("Full rule name. This field is for information purposes and can be used to create additional custom checks that reuse the same data quality rule.")
    private String ruleName;

    /**
     * Rule parameters specification, returned for reference and debugging. Used by the document generation utilities.
     */
    @JsonIgnore
    private AbstractRuleParametersSpec ruleParametersSpec;

    /**
     * List of fields for editing the rule parameters like thresholds.
     */
    @JsonPropertyDescription("List of fields for editing the rule parameters like thresholds.")
    private List<FieldModel> ruleParameters;

    /**
     * Disable the rule. The rule will not be evaluated. The sensor will also not be executed if it has no enabled rules.
     */
    @JsonPropertyDescription("Disable the rule. The rule will not be evaluated. The sensor will also not be executed if it has no enabled rules.")
    private boolean disabled;

    /**
     * Returns true when the rule is configured (is not null), so it should be shown in the UI as configured (having values).
     */
    @JsonPropertyDescription("Returns true when the rule is configured (is not null), so it should be shown in the UI as configured (having values).")
    private boolean configured;

    /**
     * Creates a selective deep/shallow clone of the object. Definition objects are not cloned, but all other editable objects are.
     * @return Cloned instance.
     */
    public RuleParametersModel cloneForUpdate() {
        try {
            RuleParametersModel cloned = (RuleParametersModel)super.clone();
            if (cloned.ruleParametersSpec != null) {
                cloned.ruleParametersSpec = (AbstractRuleParametersSpec) cloned.ruleParametersSpec.deepClone();
            }

            if (cloned.ruleParameters != null) {
                cloned.ruleParameters = cloned.ruleParameters
                        .stream()
                        .map(fieldModel -> fieldModel.cloneForUpdate())
                        .collect(Collectors.toList());
            }

            return cloned;
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported: " + ex.toString(), ex);
        }
    }

    /**
     * Applies sample values for fields that have a sample value. Overrides the current values.
     * The model filled with sample values is used for generating the documentation model.
     */
    public void applySampleValues() {
        if (this.ruleParameters != null) {
            for (FieldModel ruleParameterFieldModel : this.ruleParameters) {
                ruleParameterFieldModel.applySampleValues();
            }
        }
    }
}
